Ability to reset to remote commit

estib-vega opened the pull request:
We were able to hard reset to whatever was on the remote, in the case that e.g. one would amend a commit outside of GB and wanted to pull that in.
This re-adds that.

### Report
This was reported in Discord: https://discord.com/channels/1060193121130000425/1073202153163857920/1306629953848213585

### Changes
Adds a button to reset your local branch to whatever is on remote.
This works in the case in which there are divergent commits with the same change ID.`

### Visuals
<img width="394" alt="Screenshot 2024-11-14 at 17 48 17" src="https://github.com/user-attachments/assets/eead3d08-68e0-4f9d-a141-2ba1694e8779">

estib-vega committed:
Ability to reset to remote commit

Adds a button to reset your local branch to whatever is on remote.
This works in the case in which there are divergent commits with the same change ID.

mtsgrd commented on apps/desktop/src/lib/commits/utils.ts:
```diff
@@ -0,0 +1,15 @@
+ */
+export function findLastDivergentCommit(commits: DetailedCommit[]): DetailedCommit | undefined {
+	for (let i = commits.length - 1; i >= 0; i--) {
+		const commit = commits[i];
+		if (!commit) continue;
+		if (commit.id !== commit.remoteCommitId) {
+			return commit;
+		}
```
I'm assuming the `if (!commit) ..` line makes eslint happy because of `noUncheckedIndexedAccess`, but imho we shouldn't need to do this. One solution could be using `const commit = commits[i] as DetailedCommit`.

estib-vega replied:
Yeah that's why haha. I don't mind the paranoid extra check, but I can remove it if you prefer 👍.

Also, a more concise alternative:
```ts
if (commit!.id !== commit!.remoteCommitId) {
    return commit;
}
```

estib-vega commented:
Will wait for some input from Pavel, so that we also decide which button to show in the case that:
**There are remote-only commits** AND **there are diverged commits**. 

We probably just want to show 1 button as it will be the same thing

PavelLaptev committed:
added right padding to the commit details

PavelLaptev committed:
ui fix: upstream commits box overflow and border of the last commit

estib-vega committed:
ui fix: Line is dashed if the reset button is at the bottom

estib-vega committed:
Ability to hard-reset commits to what's on the remote

Set a integration strategy, by choosing:
- Rebase
- Merge
- Hard Reset

Add tests

estib-vega committed:
Series integration: Set the correct integration strategy 

Set the right strategy when reseting or pulling new changes into a series

estib-vega commented:
@PavelLaptev Ready for design review again :)
<img width="382" alt="Screenshot 2024-11-15 at 11 55 41" src="https://github.com/user-attachments/assets/9a4d3aef-e4fa-4085-9e7b-10ed5d0b5c04">
<img width="403" alt="Screenshot 2024-11-15 at 11 55 51" src="https://github.com/user-attachments/assets/a901575d-3774-4988-9136-25a9ed922a9a">

PavelLaptev committed:
refactor `CommitAction` component

PavelLaptev committed:
fix cards overflow the box + remove unused code

PavelLaptev committed:
fix empty branch paddings

PavelLaptev committed:
update `reset to remote…` button style

PavelLaptev committed:
update `reset to remote` button style

PavelLaptev committed:
increase the text size in the modal

estib-vega committed:
fix: be explicit about the branch integration call

The integration call needs to be explicitly called without arguments
